snow/engine/snowman/event/queue_test.go

```diff
@@ -0,0 +1,294 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package event
+
+import (
+	"context"
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/ava-labs/avalanchego/utils/set"
+)
+
+var (
+	errDuplicateInvocation  = errors.New("job already handled")
+	errUnexpectedInvocation = errors.New("job handled unexpectedly")
+)
+
+type testJob struct {
+	execute func(context.Context) error
+	cancel  func(context.Context) error
+}
+
+func (j *testJob) Execute(ctx context.Context) error {
+	return j.execute(ctx)
+}
+
+func (j *testJob) Cancel(ctx context.Context) error {
+	return j.cancel(ctx)
+}
+
+func newQueueWithJob[T comparable](t *testing.T, job Job, shouldCancel bool, dependencies ...T) *Queue[T] {
+	q := NewQueue[T]()
+	require.NoError(t, q.Register(context.Background(), job, dependencies...))
+	if shouldCancel {
+		for _, jobs := range q.jobs {
+			for _, j := range jobs {
+				j.shouldCancel = true
+			}
+		}
+	}
+	return q
+}
+
+func TestQueue_Register(t *testing.T) {
+	var calledExecute bool
```

Comment:
> In the process, maybe avoid reuse across test cases to allow use of calledExecute as a struct property instead of reuse/reset across tests.

I managed to simplify the flow, but we use the `userJob` to specify the structure of the tests... And it wasn't immediately clear to me how best to avoid that re-use with how we are handling the `wantScheduler`.